src/Admin/AbstractAdmin.php

```diff
@@ -2461,9 +2466,11 @@ public function getTranslationDomain()
      *
      * NEXT_MAJOR: remove this method
      *
+     * @param LegacyTranslatorInterface|TranslatorInterface $translator
+     *
      * @deprecated since sonata-project/admin-bundle 3.9, to be removed with 4.0
      */
-    public function setTranslator(TranslatorInterface $translator)
+    public function setTranslator(object $translator)
```

Comment:
Would not be this BC-break for someone extending the `AbstractAdmin` and also implemented this method?

Reply:
https://sandbox.onlinephpfunctions.com/code/33e12ee18a6acc787333047f632e5dd6a0367aef

It is possible to change type hint to more restricted, anyway you will get warnings. 
Change type to less restricted will throw Error.

Technicly this change will be BC-break but will work correct in production. I think upgrade note will be enought.

Reply:
@sonata-project/contributors WDYT?

Reply:
> but will work correct in production

that is not necessarily correct. On some of my projects even warnings will throw exceptions on production. Depends how you configure it.

So I think indeed this is a BC break :cry: good catch @franmomu 

How can we avoid it? :confused:

Reply:
Some idea: What if we rely on `__call` for it?

Some quick prototype 

```php
public function __call($name, $arguments)
{
    if ($name !== 'setTranslator') {
        throw new \BadMethodCallException('...');
    }

    @trigger_error(
        'The AbstractAdmin::setTranslator method is deprecated since version 3.9 and will be removed in 4.0.',
        E_USER_DEPRECATED
    );

    if (!is_array($arguments)
        || count($arguments) === 0
        || !($translator = $arguments[0]) instanceof LegacyTranslatorInterface && !$translator instanceof TranslatorInterface
    ) {
        throw new \TypeError(sprintf(
            'Argument 1 passed to "%s()" must be an instance of "%s" or "%s", instance of "%s" given.',
            __METHOD__,
            LegacyTranslatorInterface::class,
            TranslatorInterface::class,
            '...'
        ));
    }

    $this->translator = $translator;
}
```

This means we could remove the `setTranslator` and handle it like this?

Reply:
When I see this, or some other PR: https://github.com/sonata-project/SonataAdminBundle/pull/6362#discussion_r481261465.

I wonder if we should not focus on 4.0 instead. 

Either we're introducing BC-break in a minor update, which is not a good practice, either we're adding complexity to the code...

Reply:
Agree, there are more and more BC-breaks, we should not be afraid to move those to the next major for any reason. If we are afraid, it means we don't release majors often enough, and that's the issue we should fix, maybe by releasing one major per year at a given date.

Reply:
```php
namespace Symfony\Component\Translation;

if (!interface_exists('Symfony\Component\Translation\TranslatorInterface')) {
    /**
     * @deprecated since sonata-project/admin-bundle 3.x, to be removed in 4.0. Use \Symfony\Contracts\Translation\TranslatorInterface instead.
     */
    class_alias('Symfony\Contracts\Translation\TranslatorInterface', 'Symfony\Component\Translation\TranslatorInterface');
}
```

WDYT about this solution?

Based on: https://github.com/sonata-project/SonataBlockBundle/blob/3.x/src/Resources/stubs/symfony2.php

Reply:
IMO Add support for Symfony5 in Sonata3 will be awesome. Anyway force user to upgrade Sonata to v4 if they want use Symfony5 is also good thing. 

One important question: How much work with BC-break must be done in Sonata3 to support Symfony5?

Reply:
I agree about focusing on 4.0, but TBH I have no idea how far we are from having a stable, battle-tested release of 4.0 and some of the changes to be compatible with Symfony 5 have to be done anyway, so IMO I'd support for Symfony 5 in `3.x`. 

In this particular case, apparently (https://github.com/sonata-project/SonataAdminBundle/pull/6213#issuecomment-660716071) we have the same problem in 4.x (since the method hasn't been removed).

Reply:
> I agree about focusing on 4.0, but TBH I have no idea how far we are from having a stable, battle-tested release of 4.0 

We will never know if we don't start working actively on it.

We started with @jordisala1991, and the sonata 4 just need
- To remove some `NEXT_MAJOR`
- To add typehint every where
- To fix minor inconsistency in Interface.

And since https://github.com/sonata-project/SonataDoctrineORMAdminBundle/pull/1073 we can start to make real test with master branch.

> so IMO I'd support for Symfony 5 in `3.x`.

But this introduce BC-break. Or will add a lot of complexity to our code which lead to difficulty to maintain/update.